torch_xla/csrc/xla_sharding_util.cpp

```diff
@@ -382,4 +383,52 @@ std::vector<at::Tensor> ShardingUtil::ShardTensor(
   return shards;
 }
 
+void ShardingUtil::PrepareOutputShardingPropagation(
+    std::vector<XLATensorPtr>* tensors, absl::Span<const size_t> indices,
+    ComputationPtr computation,
+    std::vector<torch::lazy::BackendDataPtr>* data_placeholders,
+    std::vector<XLATensor::ShardingSpecPtr>* sharding_specs) {
+  auto computation_proto = computation->computation().proto();
+  auto tuple_shardings =
+      computation_proto.spmd_output_sharding().tuple_shardings();
+  if (tuple_shardings.empty()) {
+    return;
+  }
+  XLA_CHECK(indices.size() == tuple_shardings.size())
+      << "Expected " << indices.size() << " sharding annotations, but got "
+      << tuple_shardings.size();
+
+  for (int i = 0; i < indices.size(); ++i) {
+    auto xtensor = (*tensors)[indices[i]];
+    if (tuple_shardings[i].type()) {
+      (*sharding_specs)[i] = std::make_shared<XLATensor::ShardingSpec>(
+          tuple_shardings[i],
+          MakeShapeWithDeviceLayout(
+              xtensor->shape().get(),
+              static_cast<XlaDeviceType>(xtensor->GetDevice().type())));
+      xtensor->SetShardingSpec(*(*sharding_specs)[i]);
+
+      // Create sharded data placeholder, this will be used to
+      // hold the corresponding computation results.
+      auto sharded_data_placeholder =
+          WrapXlaData(xla::ComputationClient::Get()->WrapDataShards(
+              {}, GetVirtualDevice().toString(),
+              (*sharding_specs)[i]->shape.value(),
+              (*sharding_specs)[i]->sharding));
+
+      // Register the sharded data placeholder to the tensor and its node.
+      (*data_placeholders)[i] = sharded_data_placeholder;
+      xtensor->data()->handle = (*data_placeholders)[i];
+      if (auto ir_value = xtensor->CurrentIrValue()) {
+        if (DeviceData* device_data_node =
+                DeviceData::Cast(ir_value.node.get())) {
+          device_data_node->Assign(xtensor->data()->handle);
+        }
+      }
+    } else {
+      // Clear sharding if the output parameter is no longer sharded.
+      xtensor->ClearShardingSpec();
```

Comment:
in what case output tensor will not be sharded? I thought under SPMD it is either sharded or replicated.

Reply:
Correct. When it's replicated, it's no longer sharded. If user decides to unshard or the compiler decides to remove the propagation, then we that needs to be reflected on the XLATensor's as source of the truth. This is a safeguard for the edge use cases.

Reply:
oh ok, when you say it is not `sharded`, I thought it means it does not has sharding, hence it won't even be replicated. I guess comment here can be more specified about the replication.

Reply:
Great, I am also adding the implicitly replicated PjRtShardedData here after rebasing.